xtask/src/dependencies.rs

```diff
@@ -0,0 +1,102 @@
+use std::{collections::HashMap, time::Instant};
+
+use crate::{
+    endgroup, group,
+    logging::init_logger,
+    utils::{
+        ensure_cargo_command_is_installed, format_duration, is_current_toolchain_nightly,
+        run_cargo, Params,
+    },
+};
+
+#[derive(clap::ValueEnum, Default, Copy, Clone, PartialEq, Eq)]
+pub(crate) enum DependencyCheckType {
+    /// Run all dependency checks.
+    #[default]
+    All,
+    /// Perform an audit of all dependencies using the cargo-audit crate `<https://crates.io/crates/cargo-audit>`
+    Audit,
+    /// Run cargo-deny check `<https://crates.io/crates/cargo-deny>`
+    Deny,
+    /// Run cargo-udeps to find unused dependencies `<https://crates.io/crates/cargo-udeps>`
+    Unused,
+}
+
+pub(crate) fn run(dependency_check: DependencyCheckType) -> anyhow::Result<()> {
+    // Setup logger
+    init_logger().init();
+    // Start time measurement
+    let start = Instant::now();
+    match dependency_check {
+        DependencyCheckType::Audit => cargo_audit(),
+        DependencyCheckType::Deny => cargo_deny(),
+        DependencyCheckType::Unused => cargo_udeps(),
+        DependencyCheckType::All => {
+            cargo_audit();
+            cargo_deny();
+            cargo_udeps();
+        }
+    }
+
+    // Stop time measurement
+    //
+    // Compute runtime duration
+    let duration = start.elapsed();
+
+    // Print duration
+    info!(
+        "\x1B[32;1mTime elapsed for the current execution: {}\x1B[0m",
+        format_duration(&duration)
+    );
+
+    Ok(())
+}
+
+/// Run cargo-audit
+fn cargo_audit() {
+    ensure_cargo_command_is_installed("cargo-audit");
+    // Run cargo audit
+    group!("Cargo: run audit checks");
+    run_cargo(
+        "audit",
+        Params::from([]),
+        HashMap::new(),
+        "Cargo audit should be installed and it should correctly run",
+    );
+    endgroup!();
+}
+
+/// Run cargo-deny
+fn cargo_deny() {
+    ensure_cargo_command_is_installed("cargo-deny");
+    // Run cargo deny
+    group!("Cargo: run deny checks");
+    run_cargo(
+        "deny",
+        Params::from(["check"]),
+        HashMap::new(),
+        "Cargo deny should be installed and it should correctly run",
+    );
+    endgroup!();
+}
+
+/// Run cargo-udeps
+fn cargo_udeps() {
+    if is_current_toolchain_nightly() {
```

Comment:
Is there a way to install automatically the toolchain? Maybe we can use `cargo +nightly` instead of `cargo`? Not sure if it works @Luni-4 ?

Reply:
My original solution installed the toolchain and even updated it but I then encountered consistency issue where the wrong toolchain was selected for nested cargo or even rustup process calls. So I switched to a design where the user explicitly set the toolchain on the command line. For commands that requires a nightly toolchain, an error is triggered if the user defines a non-nightly toolchain and a message explains what to do to fix the error.